conda_build/environ.py

```diff
@@ -144,6 +144,7 @@ def get_dict(m=None, prefix=None):
         d['PKG_NAME'] = m.name()
         d['PKG_VERSION'] = m.version()
         d['PKG_BUILDNUM'] = str(m.build_number())
+        d['PKG_BUILD_STRING'] = str(m.build_id())
```

Comment:
Oh, that's an interesting idea.  It turns out this isn't the right function to do it (the output of `environ.get_dict()` must be strings only), but we could add these two lines to `jinja_context.context_processor()`:

```python
    if initial_metadata:
        ctx['metadata'] = initial_metadata
```

...which does the trick.  Well, we also need one more minor change needed in our special `UndefinedNeverFail` class:

```python
class UndefinedNeverFail(jinja2.Undefined):

    # ...
    
    def __getattr__(self, k):
        try:
            return object.__getattr__(k)
        except AttributeError:
            return UndefinedNeverFail(name=self._undefined_name + '.' + k)
```

...and then your idea works perfectly.  For example, this recipe does what you expect:

```yaml
package:
  name: conda-build-test-jinja-vars
  version: 1.0

build:
  number: 123
  string: {{CONDA_TEST_VAR}}_{{PKG_BUILDNUM}}

  script_env:
    - CONDA_TEST_VAR
  
about:
  summary: {{ metadata.name() }} is just a test package.
```

I'm intrigued by this idea, but I'm also wary of it.  This change would encourage recipe authors to depend explicitly on imlementation details within `conda-build`. 

At the moment, recipes are limited using to a predefined, documented set of variables (e.g. `PKG_BUILDNUM`).  No matter what changes we make to `conda-build` in the future, it will be easy to avoid breaking backward compatibility with old recipes.

But if we expose the `MetaData` class directly to the recipe, then simple changes to `conda-build` (say, renaming `build_id()` to `build_string()`) have the potential to break recipes in the wild, if those recipes use `{{ metadata.build_id() }}`.

In conclusion: I tried your idea on my machine, and it works.  But after some thought, I don't want to add it to this PR.